### PR TITLE
Allow per-species definition of KPP absolute and relative solver tolerances; Set absolute tolerances of dummy species to large value

### LINKED ISSUE #1753: Assign a very high absolute tolerance to prod/loss family species in KPP-built mechanisms
### Name and Institution (Required)

Name: Bob Yantosca
Institution: Harvard + GCST

### Confirm you have reviewed the following documentation

- [x] [Contributing guidelines](https://geos-chem.readthedocs.io/en/stable/reference/CONTRIBUTING.html)

### New GEOS-Chem feature or discussion

On the [KPP repository](https://github.com/KineticPreProcessor/KPP) there has been some discussion into preventing prod/loss species from affecting the error norm.   One recommendation from @obin1 is to set the absolute tolerance ATOL of prod/loss species to an extremely high value.:

> Here's a visualization from a project @msl3v and I've been working on: giving prod/loss species arbitrarily large absolute tolerances of 1e25 (effectively removing them from the error norm) reduces the number of integration steps. This can also slightly alter the concentrations of species whose tolerances we don't modify, by changing the internal timesteps. visualizePL

Because the error norm is used to compute KPP internal integration timesteps, this could reduce the amount of time spent in the integrator.  As time allows, we might want to try implementing this recommendation in GEOS-Chem. 

Tagging @msl3v @obin1 @RolfSander @msulprizio @jimmielin @lizziel @christophkeller for reference.

### PR DESCRIPTION
**Pull request by yantosca**
### Name and Institution (Required)

Name: Bob Yantosca
Institution: Harvard + GCST

### Describe the update
This is the companion PR to #1753.  We have implemented the following:

1. Now allow for the definition of KPP absolute tolerance (`KPP_AbsTol`)  and/or relative tolerance (`KPP_RelTol`) for each species in the `species_database.yml` file

2. Added `KPP_AbsTol` and `KPP_RelTol` tags to the `Species` defined type.  This is the type that defines the species database.

3. Updated `Headers/species_database_mod.F90` to parse the `KPP_AbsTol` and `KPP_RelTol` tags from the `species_database.yml` file.  A misisng value will be assigned to species that do not have these tags.

4. Added `State_Chm%KPP_AbsTol` and `State_Chm%KPP_RelTol` arrays to hold the KPP absolute and relative tolerances in a common place.

5. Replace missing values in `State_Chm%KPP_AbsTol` and `State_Chm%KPP_RelTol` during the initialization of the fullchem and mercury mechanisms.  This step only has to be done once, which will save excessive CPU cycles.

6. On each chemistry timestep, copy the values of `State_Chm%KPP_AbsTol` to the KPP `ATOL` array.  Also copy the values of `State_Chm%KPP_RelTol` to the KPP `RTOL` array.

7. Edit the `run/shared/species_database.yml` and `run/shared/species_database_hg.yml` template files so that dummy species have a `KPP_AbsTol` value of 1.0e25.  This will prevent these species from influencing the Rosenbrock solver error norm computation.

### Expected changes
1. This will allow per-species setting of KPP tolerances, which may help to prevent convergence errors.  It will also facilitate a parameter exploration of tolerances by interested users.
2. Integration of the chemical mechanism should be more accurate (and maybe take less time to converge to a solution) due to the removal of the dummy species from the error norm computation.

### Related Github Issue

- Closes #1753
- Also see https://github.com/KineticPreProcessor/KPP/issues/66
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

### CODE REVIEW

**Comment by yantosca** [comment]
Also tagging @obin1 @jimmiel

File: run/shared/species_database.yml [hunk 1]
```diff
@@ -2760,18 +2760,22 @@ LBRO2H:
   FullName: Dummy species to track oxidation of BRO2 by HO2
   Is_Gas: true
   MW_g: 159.13
+  KPP_AbsTol: 1.0e+25_8
```
Comment (msulprizio) [inline run/shared/species_database.yml:2763]: Fix typo

**Comment by yantosca** [comment]
Fixing a couple of bugs.  Will push a fix shortly.

**Review by msulprizio — CHANGES_REQUESTED** [review]
Thanks @yantosca. These changes look good. I just found a few typos to fix.

Have you run any simulations with these changes yet? I'm curious if it's a zero difference change as of now or if we see some changes on the level of numerical noise. It would be nice to note that in this PR. Additionally, it would be good to see some stats on how this impacts model performance.

**Comment by yantosca** [comment]
All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

GCHP      #966d10d GCHP 14.4.1 release
GEOS-Chem #813ea4332 Fixed typo in fullchem_mod.F90
HEMCO     #

Number of execution tests: 11

Submitted as SLURM job: 38380509
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Comment by yantosca** [comment]
@msulprizio, I haven't run some simulations yet but I could set up a couple of quick ones.  My guess is that this will not be a zero-diff update, which is why I have it going into 14.5.0.  I'll post it here.

**Review by msulprizio — APPROVED** [review]
Thanks. Once confirmed the changes are zero or minor then I approve the merge. Could you also post timers of ref and dev runs when you have them?

**Comment by yantosca** [comment]
@msulprizio: I ran 3-day simulations for 14.4.1 and this PR.  Both ran on the same hardware (huce_cascade, 48 cores).  Excluding the dummy species from the KPP error norm computations (i.e. setting them to ATOL = 1e25) resulted in a simulation that was 8% faster overall and 4% faster in gas-phase chemistry.
```console
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% GEOS-Chem Classic Benchmark Timing Information
%%%
%%% Ref = 14.4.1
%%% Dev = PR_#2359
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


Timer                              Ref [s]             Dev [s]         % Diff
-------------------------------------------------------------------------------
GEOS-Chem                         1456.250            1338.875     -8.060e+00
HEMCO                              443.750             343.000     -2.270e+01
All chemistry                      434.000             423.125     -2.506e+00
=> Gas-phase chem                  272.875             259.750     -4.810e+00
=> Photolysis                       41.125              43.375      5.471e+00
=> Aerosol chem                    104.250             103.375     -8.393e-01
=> Linearized chem                   2.125               2.125      0.000e+00
Transport                           82.375              85.125      3.338e+00
Convection                         100.750             101.875      1.117e+00
Boundary layer mixing              135.000             133.375     -1.204e+00
Dry deposition                       2.750               3.000      9.091e+00
Wet deposition                      34.250              37.375      9.124e+00
Diagnostics                        160.625             160.875      1.556e-01
Unit conversions                    38.375              37.000     -3.583e+00
```
Also all of the differences were at the numerical noise level.  Below is output from the `compare_diags.py` script in GCPy.  The sum of Ref, sum of Dev, and diff is printed.  The diffs are 1e-5 or 1e-6 below the order of magnitude of the concentrations, which indicates numerical noise.

(gcpy_env) [holy8a24302 runtests]$ python -m gcpy.examples.diagnostics.compare_diags
Using configuration file compare_diags.yml
... Creating single-level plots
... Creating zonal mean plots
... Printing totals and differences
... Only showing variables with |absolute difference| > 0.0
Variable               Ref=14.4.1               Dev=PR #2359             Dev - Ref
SpeciesConcVV_A3O2   : 1.0102523e-09          | 1.0102343e-09          | -1.807603715417872e-14 
SpeciesConcVV_ACET   : 7.050497e-05           | 7.050495e-05           | -1.517309715370896e-11 
SpeciesConcVV_ACTA   : 5.8778414e-06          | 5.8778487e-06          | 7.31086458732877e-12 
SpeciesConcVV_AERI   : 7.796375e-08           | 7.796357e-08           | -1.8312192311565473e-13 
SpeciesConcVV_ALD2   : 5.4596476e-06          | 5.459645e-06           | -2.666621264080238e-12 
SpeciesConcVV_ALK4   : 6.014887e-06           | 6.0148923e-06          | 5.128444110352737e-12 
SpeciesConcVV_AONITA : 3.1718568e-07          | 3.1718474e-07          | -9.431420488342779e-13 
SpeciesConcVV_AROMP4 : 2.550492e-09           | 2.5504927e-09          | 3.43391416797048e-18 
SpeciesConcVV_AROMP5 : 9.744214e-10           | 9.744252e-10           | 3.811899831429686e-15 
SpeciesConcVV_AROMRO2 : 5.585208e-10           | 5.5853305e-10          | 1.2144287447413485e-14 
SpeciesConcVV_ASOA1  : 4.181633e-08           | 4.181629e-08           | -4.284428172483812e-14 
SpeciesConcVV_ASOA2  : 1.2599112e-08          | 1.2599083e-08          | -2.8944481224790417e-14 
SpeciesConcVV_ASOA3  : 3.840629e-08           | 3.8406256e-08          | -3.3075447356022483e-14 
SpeciesConcVV_ASOAN  : 3.0354474e-07          | 3.0354505e-07          | 3.1313558516987694e-13 
SpeciesConcVV_ASOG1  : 8.0630706e-08          | 8.063071e-08           | 4.4828765570807064e-15 
SpeciesConcVV_ASOG2  : 5.2565404e-08          | 5.2565362e-08          | -3.990944808787353e-14 
SpeciesConcVV_ASOG3  : 1.1648519e-06          | 1.1648519e-06          | -3.608927528564801e-14 
SpeciesConcVV_ATO2   : 4.0639825e-09          | 4.0640202e-09          | 3.8777463092767514e-14 
SpeciesConcVV_ATOOH  : 1.2774818e-06          | 1.2774756e-06          | -6.4869021612623534e-12 
SpeciesConcVV_B3O2   : 4.0574264e-09          | 4.0573442e-09          | -8.117796392774837e-14 
SpeciesConcVV_BALD   : 7.3409057e-09          | 7.3408857e-09          | -2.009821481333343e-14 
SpeciesConcVV_BENZ   : 1.3626452e-06          | 1.3626454e-06          | 2.1519315192633504e-13 
SpeciesConcVV_BENZO  : 1.1949922e-10          | 1.1949938e-10          | 1.6769407411996908e-16 
SpeciesConcVV_BENZO2 : 2.221213e-09           | 2.2212292e-09          | 1.656984486143702e-14 
SpeciesConcVV_BENZP  : 3.8712625e-07          | 3.8712705e-07          | 7.850934529221587e-13 
SpeciesConcVV_BRO2   : 3.0105338e-10          | 3.010559e-10           | 2.4890986420980123e-15 
SpeciesConcVV_BUTDI  : 5.436187e-05           | 5.436187e-05           | -7.480110281177232e-13 
SpeciesConcVV_BZCO3  : 8.6094725e-12          | 8.610297e-12           | 8.248203552455036e-16 
SpeciesConcVV_BZCO3H : 2.5661777e-09          | 2.5661988e-09          | 2.1030625293539176e-14 
SpeciesConcVV_BZPAN  : 5.422563e-09           | 5.422623e-09           | 5.957895580967904e-14 
SpeciesConcVV_Br     : 6.5181945e-07          | 6.5181956e-07          | 1.534886990726611e-13 
SpeciesConcVV_Br2    : 1.3201482e-07          | 1.3201436e-07          | -4.402653914180321e-13 
SpeciesConcVV_BrCl   : 3.2360273e-07          | 3.236069e-07           | 4.211647090107684e-12 
SpeciesConcVV_BrNO2  : 4.039457e-09           | 4.0395527e-09          | 9.60652456021005e-14 
SpeciesConcVV_BrNO3  : 4.5599748e-07          | 4.5599907e-07          | 1.5647613413327655e-12 
SpeciesConcVV_BrO    : 9.739696e-07           | 9.739707e-07           | 1.0496989562291947e-12 
SpeciesConcVV_BrSALA : 1.638811e-08           | 1.6388142e-08          | 3.200783347677601e-14 
SpeciesConcVV_BrSALC : 5.739084e-08           | 5.7390622e-08          | -2.2099898679735913e-13 
SpeciesConcVV_C2H2   : 3.895558e-06           | 3.8955577e-06          | -3.432070708360435e-13 
SpeciesConcVV_C2H4   : 4.4643784e-06          | 4.464386e-06           | 6.87820893990887e-12 
SpeciesConcVV_C2H6   : 3.158488e-05           | 3.1584874e-05          | -5.373543623954369e-12 
SpeciesConcVV_C3H8   : 4.4658354e-06          | 4.465839e-06           | 3.2593784442258578e-12 
SpeciesConcVV_C4HVP1 : 8.47547e-12            | 8.475108e-12           | -3.6249132319772406e-16 
SpeciesConcVV_C4HVP2 : 2.7173411e-11          | 2.7171959e-11          | -1.4490189452305764e-15 
... etc ...
```

**Comment by yantosca** [comment]
@msulprizio: Here is the difference plot for O3.  Other species show similar numerical noise differences.
o3

**Comment by obin1** [comment]
That's a similar timing speed up we've seen with GEOS-CF, and concentration differences are also similar to what we saw with some KPP Standalone experiments: https://github.com/KineticPreProcessor/KPP/issues/66#issuecomment-1508762259

Interestingly, in those experiments, we saw that the species most affected (still on the order of numerical noise) weren't the P/L species whose tolerances were adjusted. I think comes from the sensitivity of the solution number of internal steps taken

**Comment by yantosca** [comment]
Thanks for confirming @obin1.  I'll start the process to merge this into the `dev/14.5.0` branch.

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #6f7379d GCClassic (GEOS-Chem Classic) 14.4.1 release
GEOS-Chem #4288f0d1d Merge PR #2359 (Set KPP dummy species to large ATOL value)
HEMCO     #2192e0e HEMCO 3.9.1 release

Using 24 OpenMP threads
Number of execution tests: 28

Submitted as SLURM job: 38552013
==============================================================================
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

All GCHP integration tests passed as well:
```console
==============================================================================
GCHP: Execution Test Results

GCHP      #966d10d GCHP 14.4.1 release
GEOS-Chem #4288f0d1d Merge PR #2359 (Set KPP dummy species to large ATOL value)
HEMCO     #

Number of execution tests: 11

Submitted as SLURM job: 38552345
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Comment by msulprizio** [comment]
Thanks for posting those results @yantosca!

**Comment by yantosca** [comment]
Note: 1-hr GCHP cloud benchmarks on this PR failed due to the error described in PR #2363.  One PR #2363 is approved I will merge that into this PR.

**Pull request merged**